tests/testthat/test-join_keys.R

```diff
@@ -0,0 +1,666 @@
+  )
+})
+
+testthat::test_that("join_keys doesn't accept other objects than teal_data, TealData and join_key", {
+  testthat::expect_error(join_keys("a")) # todo: add expected error message
+})
+
+testthat::test_that("join_keys doesn't accept a list which is identical to output of join_key function", {
+  key <- join_key("a", "b", "test")
+  testthat::expect_error(join_keys(unclass(key)))
+})
+
+testthat::test_that("join_keys fails when provided foreign key pairs have incompatible values", {
+  testthat::expect_error(
+    join_keys(join_key("d1", "d2", "a"), join_key("d2", "d1", "b")),
+    "cannot specify multiple different join keys between datasets"
+  )
+  testthat::expect_error(
+    join_keys(join_key("d1", "d2", c(a = "b")), join_key("d2", "d1", c(a = "b"))),
+    "cannot specify multiple different join keys between datasets"
+  )
+
+  testthat::expect_error(
+    join_keys(
+      join_keys(
+        join_key("q", "b", "d"),
+        join_key("a", "b", "c")
+      ),
+      join_key("a", "q", "e"),
+      join_key("a", "b", "f")
+    ),
+    "cannot specify multiple different join keys between datasets"
+  )
+})
+
+testthat::test_that("join_keys constructor adds symmetric keys on given (unnamed) foreign key", {
+  my_keys <- join_keys(join_key("d1", "d2", "a"))
+  testthat::expect_identical(
+    my_keys,
+    join_keys(join_key("d1", "d2", "a"), join_key("d2", "d1", "a"))
+  )
+})
+
+testthat::test_that("join_keys constructor adds symmetric keys on given (named) foreign key", {
+  testthat::expect_identical(
+    join_keys(
+      join_key("d1", "d2", c(a = "b"))
+    ),
+    join_keys(
+      join_key("d1", "d2", c(a = "b")),
+      join_key("d2", "d1", c(b = "a"))
+    )
+  )
+})
+# [.join_keys -----------------------------------------------------------------
+testthat::test_that("[[.join_keys returns keys for given pair", {
+  my_keys <- join_keys(
+    join_key("a", "a", "aa"),
+    join_key("b", "b", "bb"),
+    join_key("c", "c", "cc"),
+    join_key("b", "a", "child-parent"),
+    join_key("c", "a", "child-parent")
+  )
+  testthat::expect_identical(my_keys[["b"]][["a"]], c(`child-parent` = "child-parent"))
+})
+
+testthat::test_that("[[.join_keys doesn't return keys for given a pair without explicit join_key", {
+  my_keys <- join_keys(
+    join_key("a", "a", "aa"),
+    join_key("b", "b", "bb"),
+    join_key("c", "c", "cc"),
+    join_key("b", "a", "child-parent"),
+    join_key("c", "a", "child-parent")
+  )
+  testthat::expect_null(my_keys[["b"]][["c"]])
+})
+
+testthat::test_that("[[.join_keys infer keys between child by shared foreign keys to parent ", {
+  my_keys <- join_keys(
+    join_key("a", "a", "aa"),
+    join_key("b", "b", "bb"),
+    join_key("c", "c", "cc"),
+    join_key("b", "a", "child-parent"),
+    join_key("c", "a", "child-parent")
+  )
+  parents(my_keys) <- list("b" = "a", "c" = "a")
+  testthat::expect_identical(my_keys[["b"]][["c"]], c(`child-parent` = "child-parent"))
```

Comment:
Shouldn't the result be `"aa"`?

I believe the logic of `update_keys_given_parents` sets the "implicit key" as the parent's primary key

Reply:
Hmmm, i think childs should be connected on the same keys they are connected to parent. Parent primary might not exist in childs. For example

```r
parent (id) -> (id) parent
child1 (parent_id) -> (id) parent 
child2 (parent_id) -> (id) panent
 ```

Reply:
Ok, it works now this way with improved logic that looks for common columns.

See extension for your new test: 
`[[.join_keys infer keys between child by shared foreign keys to parent (key names are unique to datasets)`